### ISSUE TITLE
Not uploading directories to S3

### ISSUE THREAD

**Issue by aguynamedben** [antonagestam/collectfast#123]
Hi, thanks for building this promising library. I'm looking forward to using it.

My app uses the default value for `STATICFILES_FINDERS`

```py
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
)
```

When asset are collected into static/, I have these files:

```
$ ls -alh static
total 4200
drwxr-xr-x  16 ben  staff   512B Jan 24 15:49 .
drwxr-xr-x  61 ben  staff   1.9K Jan 25 10:59 ..
-rw-r--r--   1 ben  staff   669B Jan 24 15:49 500.html
drwxr-xr-x   6 ben  staff   192B Jan 24 15:49 admin
drwxr-xr-x   5 ben  staff   160B Jan 24 15:49 debug_toolbar
-rw-r--r--   1 ben  staff   209B Jan 24 15:49 demo_config.yaml
-rw-r--r--   1 ben  staff   671K Jan 24 15:49 demo_deed.pdf
-rw-r--r--   1 ben  staff   101K Jan 24 15:49 demo_statement.pdf
drwxr-xr-x   5 ben  staff   160B Jan 24 15:49 django_extensions
drwxr-xr-x   5 ben  staff   160B Jan 24 15:49 gis
drwxr-xr-x  13 ben  staff   416B Jan 24 15:49 images
-rw-r--r--   1 ben  staff    23B Jan 24 15:49 robots.txt
```

There are files (500.html, demo_config.yaml) and there are directories (debug_toolbar, images, gis). When I use Collectfast, only the files are uploaded to S3.

```
$ ./manage.py collectstatic --settings=settings_local --no-input --clear
Copying '/Users/ben/code/minerals/core/static/500.html'
Copying '/Users/ben/code/minerals/core/static/robots.txt'
Copying '/Users/ben/code/minerals/core/static/demo_deed.pdf'
Copying '/Users/ben/code/minerals/core/static/demo_statement.pdf'
4 static files copied.
```

I expected the files *and* directories to be uploaded to S3, not just the files. When I use `--disable-collectfast`, causing `storages.backends.s3boto3.S3Boto3Storage` to be used directly, the files *and* directories are uploaded to S3.

```
$ ./manage.py collectstatic --settings=minerals.settings_local --no-input --clear --disable-collectfast
...(lots of copying of a bunch of files)...
226 static files copied.
```

It's surprising behavior that Collectfast doesn't upload the directories like S3Boto3Storage does when used directly (I think I also tried with S3BotoStorage). **Is there a way to get Collectfast to upload the directories?**

Here's my config...

```py
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
)

# I verified that the /var/tmp/collectfast_cache directory is being used as expected
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": "redis://127.0.0.1:6379/2",
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient",
        }
    },
    'collectfast': {
        'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
        'LOCATION': '/var/tmp/collectfast_cache',
        'TIMEOUT': None,
        'OPTIONS': {
            'MAX_ENTRIES': 1000,
        },
    }
}

# Collectfast
STATICFILES_STORAGE = "storages.backends.s3boto3.S3Boto3Storage"
AWS_PRELOAD_METADATA = True
COLLECTFAST_THREADS = 20
COLLECTFAST_DEBUG = True
COLLECTFAST_CACHE = 'collectfast'

# I've also setup Boto3 settings correctly (`AWS_STORAGE_BUCKET_NAME`, `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_QUERYSTRING_EXPIRE`)
```

Thanks for any help, and thanks for the time you spend on this library. I was surprised at how slow collectstatic was when used with ManifestStaticFilesStorage and S3Boto3Storage, so I was very excited to find this library.

**Comment by antonagestam** [comment]
I'm not able to reproduce this, see below. I tried with the `--clear` flag too (could it be an issue related to Collectfast not respecting that flag?). I'm using redis as a cache backend but that shouldn't really matter.

```
$ mkdir static/test1
$ mkdir static/test2
$ echo 'woot' > static/test1/test.txt
$ echo 'woot' > static/test2/test.txt
$ python manage.py collectstatic --no-input -v 3
Deleting 'test2/test.txt'
Deleting 'test1/test.txt'
Copying '/Users/antonagestam/python_projects/antonagestam.se/static/test2/test.txt'
Copying '/Users/antonagestam/python_projects/antonagestam.se/static/test1/test.txt'
2 static files copied.
```

What versions of Django and Python and what OS are you on?

**Comment by aguynamedben** [comment]
@antonagestam In the 1.5 days I was messing around with all of our asset management, somebody recommended [WhiteNoise](http://whitenoise.evans.io/en/stable/) to me, which was appropriate for our app, so I used that. I spent more time on re-jiggering our assets than I was supposed to, so I had to move on.

Feel free to close. Maybe somebody will run into the same thing one day and find this via Google. Thanks for the effort you put into making this public.
